File: ios/NotiSync/BrokerPairingHost.swift
import Foundation

/// Rotating app host sessions. The existing session remains usable during the final 30-second overlap.
@MainActor enum BrokerPairingHost {
    private enum Event: Sendable {
        case ready(Int, BrokerPairingLink), rotate(Int), expired(Int), finished(Int, PairingCandidate?), retry
    }

    static func awaitCard(
        exchange: @escaping @Sendable (@escaping @Sendable (BrokerPairingLink) -> Void) async throws -> PairingCandidate,
        onLink: (BrokerPairingLink?) -> Void,
        onUnavailable: () -> Void
    ) async throws -> PairingCandidate {
        let (events, continuation) = AsyncStream<Event>.makeStream()
        var sessions: [Int: Task<Void, Never>] = [:]
        var links: [Int: BrokerPairingLink] = [:]
        var expired: Set<Int> = []
        var newest = 0, displayed: Int?
        var rotationDue = false
        var retry: Task<Void, Never>?
        var retrySeconds: TimeInterval = 2

        func displayLatest() {
            let next = links.keys.max()
            if next != displayed { displayed = next; onLink(next.flatMap { links[$0] }) }
        }
        func start() {
            precondition(sessions.count < 2)
            newest += 1; rotationDue = false
            let id = newest
            sessions[id] = Task {
                let timer = Task {
                    do {
                        try await Task.sleep(for: .seconds(BrokerPairingLink.sessionSeconds - 30))
                        continuation.yield(.rotate(id))
                        try await Task.sleep(for: .seconds(30))
                        continuation.yield(.expired(id))
                    } catch { }
                }
                defer { timer.cancel() }
                let candidate = try? await exchange { continuation.yield(.ready(id, $0)) }
                continuation.yield(.finished(id, Task.isCancelled ? nil : candidate))
            }
        }
        let candidate: PairingCandidate? = await withTaskCancellationHandler {
            onLink(nil)
            start()
            for await event in events {
                if Task.isCancelled { break }
                switch event {
                case let .ready(id, link):
                    if sessions[id] != nil && !expired.contains(id) {
                        links[id] = link; retrySeconds = 2; displayLatest()
                    }
                case let .rotate(id):
                    if id == newest { rotationDue = true; if sessions.count < 2 { start() } }
                case let .expired(id):
                    if sessions[id] != nil {
                        expired.insert(id); links[id] = nil; displayLatest()
                        sessions[id]?.cancel()
                        if links.isEmpty { onUnavailable() }
                    }
                case let .finished(id, received):
                    await sessions.removeValue(forKey: id)?.value
                    links[id] = nil; expired.remove(id)
                    if let received { return received }
                    displayLatest()
                    if links.isEmpty { onUnavailable() }
                    if rotationDue && sessions[newest] != nil && sessions.count < 2 { start() }
                    else if id == newest && retry == nil {
                        let wait = retrySeconds
                        retry = Task {
                            do { try await Task.sleep(for: .seconds(wait)); continuation.yield(.retry) } catch { }
                        }
                        retrySeconds = min(retrySeconds * 2, 30)
                    }
                case .retry:
                    retry = nil
                    if sessions.count < 2 { start() }
                }
            }
            return nil
        } onCancel: { continuation.finish() }
        for task in sessions.values { task.cancel() }
        retry?.cancel()
        for task in sessions.values { await task.value }
        await retry?.value
        continuation.finish(); onLink(nil)
        try Task.checkCancellation()
        guard let candidate else { throw CancellationError() }
        return candidate
    }
}

File: peer-core/src/main/kotlin/net/extrawdw/notisync/peer/pairing/BrokerPairingHost.kt
package net.extrawdw.notisync.peer.pairing

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import net.extrawdw.notisync.protocol.BrokerPairing

/** Owns rotating broker-assisted host sessions for one pairing screen. No device is trusted here. */
class BrokerPairingHost(
    private val sessionMillis: Long = BrokerPairing.SESSION_MILLIS,
    private val rotationLeadMillis: Long = 30_000,
    private val retryMillis: Long = 2_000,
    private val exchange: suspend (onReady: (BrokerPairingLink) -> Unit) -> String,
) {
    init {
        require(rotationLeadMillis in 1 until sessionMillis && retryMillis > 0)
    }

    /** Returns only the first authenticated CARD, after closing every other session. */
    suspend fun awaitCard(
        onLink: (BrokerPairingLink?) -> Unit,
        onUnavailable: () -> Unit,
    ): String = coroutineScope {
        val events = Channel<Event>(Channel.UNLIMITED)
        val sessions = mutableMapOf<Int, Job>()
        val links = mutableMapOf<Int, BrokerPairingLink>()
        val expired = mutableSetOf<Int>()
        var newest = 0
        var displayed: Int? = null
        var retry: Job? = null
        var retryDelay = retryMillis
        var rotationDue = false

        fun displayLatest() {
            val next = links.keys.maxOrNull()
            if (next != displayed) {
                displayed = next
                onLink(next?.let(links::getValue))
            }
        }

        fun startSession() {
            check(sessions.size < 2)
            rotationDue = false
            val id = ++newest
            sessions[id] = launch {
                // Start timers before connecting, so network/setup delays cannot extend the QR lifetime.
                val rotate = launch {
                    delay(sessionMillis - rotationLeadMillis)
                    events.send(Event.Rotate(id))
                    delay(rotationLeadMillis)
                    events.send(Event.Expired(id))
                }
                var payload: String? = null
                try {
                    payload = withTimeout(sessionMillis) {
                        exchange { link -> events.trySend(Event.Ready(id, link)) }
                    }
                } catch (_: TimeoutCancellationException) {
                    // Expiry is normal; the next session may already be displayed.
                } catch (cancelled: CancellationException) {
                    throw cancelled
                } catch (_: Exception) {
                    // Authentication/network failures burn this session. A retry gets a fresh secret.
                } finally {
                    rotate.cancel()
                    events.trySend(Event.Finished(id, payload))
                }
            }
        }

        try {
            onLink(null)
            startSession()
            var received: String? = null
            while (received == null) {
                when (val event = events.receive()) {
                    is Event.Ready -> if (event.id in sessions && event.id !in expired) {
                        links[event.id] = event.link
                        retryDelay = retryMillis
                        displayLatest()
                    }
                    is Event.Rotate -> if (event.id == newest) {
                        rotationDue = true
                        if (sessions.size < 2) startSession()
                    }
                    is Event.Expired -> if (event.id in sessions) {
                        expired += event.id
                        links.remove(event.id)
                        displayLatest()
                        if (links.isEmpty()) onUnavailable()
                    }
                    is Event.Finished -> {
                        sessions.remove(event.id)?.join()
                        links.remove(event.id)
                        expired.remove(event.id)
                        if (event.payload != null) {
                            received = event.payload
                        } else {
                            displayLatest()
                            if (links.isEmpty()) onUnavailable()
                            if (rotationDue && newest in sessions && sessions.size < 2) {
                                // A closing older socket may have occupied the second slot at rotation time.
                                startSession()
                            } else if (event.id == newest && retry == null) {
                                val waitMillis = retryDelay
                                retry = launch { delay(waitMillis); events.send(Event.Retry) }
                                retryDelay = (retryDelay * 2).coerceAtMost(30_000)
                            }
                        }
                    }
                    Event.Retry -> {
                        retry = null
                        if (sessions.size < 2) startSession()
                    }
                }
            }
            received
        } finally {
            // Cancel both first, then wait for their sockets to close before returning a candidate.
            sessions.values.forEach(Job::cancel)
            retry?.cancel()
            withContext(NonCancellable) {
                sessions.values.forEach { it.join() }
                retry?.cancelAndJoin()
            }
            events.cancel()
            onLink(null)
        }
    }

    private sealed interface Event {
        data class Ready(val id: Int, val link: BrokerPairingLink) : Event
        data class Rotate(val id: Int) : Event
        data class Expired(val id: Int) : Event
        data class Finished(val id: Int, val payload: String?) : Event
        data object Retry : Event
    }
}
